### ISSUE TITLE
Log libqaul errors to file

### ISSUE THREAD

**Issue by MathJud** [qaul/qaul.net#297]
Write libqaul errors to file.

* the file shall be written into the same directory as the configuration file (config.yaml).
* only rust errors shall be logged into it.
* create `log.rs` file to handle the logic in the `storage` module https://github.com/qaul/qaul.net/tree/main/rust/libqaul/src/storage

**Comment by GildedHonour** [comment]
working on this

**Comment by GildedHonour** [comment]
@MathJud 

`log_rust_error(...)` -->


will it be a helper function which would be called manually by user? 
Or would it be called automatically whenever any rust, probably unhandled, error occurs?

**Comment by MathJud** [comment]
> @MathJud
> 
> `log_rust_error(...)` -->
> 
> will it be a helper function which would be called manually by user? Or would it be called automatically whenever any rust, probably unhandled, error occurs?

Any Error should optimally - if possible - automatically be logged to a file, for that all errors are catched.

**Comment by GildedHonour** [comment]
possible solutions:

(1)
* The library will create a new log file every day with the name `{day}{month}{year}.log`
* in case of exception it'll send the last, say, 2-3 error log files

* How would the library signal to UI that an new error has just occured?


(2)
* the name of a error log file will be `{linux_timestamp}.log`
* UI will scan the error log directory every, say, 5 minutes, possibly more frequently. If there's any file - send it and then delete it.
* or -- send an error log file and add `__sent` to its name; don't delete it immediately;  keep the last N files and delete others

**Comment by GildedHonour** [comment]
When an error occurs, the app can't simply send a report, or stacktrace, about it on its own, in background -- without notifying the user, because there may not be internet because it's a Mesh app. Therefore, the user has to be shown and asked to send an error report by himself. 

Correct? @MathJud

**Comment by MathJud** [comment]
Yes, the user needs to send it manually.

Some don't even want to log.
You check the GUI screen, where the user can send the logs via email, when you click on the three dots and select debug. There the user also can select if the application shall log at all ...

This just hits me: we should be able to configure it ... 

On a side note. An error in libqaul is usuall no problem and does not let anything to crash. There are only a few that can panick.

**Comment by GildedHonour** [comment]
Turning on/off the logger **dynamically** may not be possible because its state has to be specified during compilation, in our case during the compilation of the library. The same goes for logging levels - they have to be specified in the code, statically.



Do we really need to turn it off? Will it hurt performance when it's on? 

Or we could have 2 builds -- "normal" and "dev", the latter would be with the logger on and "log::debug" level.

We can, however, specify ENV variables which we'd pass during compilation (on/off) and thus we'd avoid changing the code. Perhaps the levels could be specified this way too.


Doing a research, may be there're some workarounds...

**Comment by GildedHonour** [comment]
https://rust-lang-nursery.github.io/rust-cookbook/development_tools/debugging/config_log.html#log-messages-to-a-custom-location


`RUST_LOG="warn,test::foo=info,test::foo::bar=debug" ./test`

===>

I guess that might work for tech-savvy users.

**Comment by GildedHonour** [comment]
Why can't we have a global variable `GLOBALS.is_logger_enabled()` which would switch from on to off and vise versa when the user checks or unchecks checkbox? This would require to wrap  every `log::debug` with `if GLOBALS.is_logger_enabled()`, but it'd work

**Comment by MathJud** [comment]
This is a sketch for a solution

1) Use log4rs for logging.
  * https://docs.rs/log4rs/latest/log4rs/
  * configuration & usage example: https://rust-lang-nursery.github.io/rust-cookbook/development_tools/debugging/config_log.html#log-messages-to-a-custom-location
2) use the configuration 
  * you'll get back a handle for later reconfiguration.
  * https://docs.rs/log4rs/1.0.0/log4rs/config/fn.init_config.html
3) reconfigure to write to file, to stop writing to file
  * you can invoke the set_config() function on the handle to change configuration.
  * https://docs.rs/log4rs/1.0.0/log4rs/struct.Handle.html#method.set_config
4) To log to a file you can use the `rolling_file_appender` feature
  * https://docs.rs/log4rs/latest/log4rs/append/rolling_file/index.html


In the logger module we should have the these public functions (dummy code)

```rs
/// Initialize logger
pub fn init() {
	// create configuration structure
	
	// start logging
	
	// by default log only errors to std_out
	
	// save handle to state
}

/// configure 
/// 
/// save state locally 
pub fn configure(log_enabled: bool, log_level: enum) {
	// create configuration structure
	
	// get handle to logger from state
	
	// set new configuration to logger
}
```

Now we can reconfigure logging whenever we want

A) after loading the confiruation file
B) when receiving a protobuf message from the GUI to set logging a certain way.


## Questions

* Does it work in parallel with android_logger?
  * logging start for android: https://github.com/qaul/qaul.net/blob/5b3786a607cf2fb6cdba765605c53d341cf458dd/rust/libqaul/src/api/android.rs#L41
  * We could use the android_logger only in the debug build, using `#[cfg(debug_assertions)]`
* If the application crashes, can we find the last error in the log file?
  * this should be tested

**Comment by GildedHonour** [comment]
Yeah, but our initial requirement is to be able to turn it on and off **dynamically**, at runtime, as well as change log levels too. That logger doesn't support that, does it?

**Comment by GildedHonour** [comment]
- "write logs" -> by default unchecked (turned off) --> logs get written into stdout / cli
- "write logs" -> when checked (turned on) ---> logs get written into a file

**Comment by GildedHonour** [comment]
- how to, if possible at all, intercept and write into a log file the latest error which makes an app crash complitely, and which is the case of a crash
===>

it's probably impossible and instead there needs to be a watchdog

**Comment by GildedHonour** [comment]
> Does it work in parallel with android_logger? 

reply from someone:


![image](https://user-images.githubusercontent.com/2032888/156881817-4ba96318-9079-41f2-adf8-8c3e05679083.png)

**Comment by GildedHonour** [comment]
Let's use `log4rs`.

This library https://github.com/tokio-rs/tracing has been suggested as well, but it's got 300 issues on github, over 1k of commits and hasn't reached 1.0 yet.

**Comment by GildedHonour** [comment]
A draft version is ready

**Comment by GildedHonour** [comment]
I figure that either log4rs or android_logger will get activated in runtime whenever `log::abc!(...)` is called

Perhaps also

```
#![not(cfg(target_os = "android"))]
```

will have to be into the files that use `log4rs`

**Comment by GildedHonour** [comment]
removed the brach alex_mff_dev1
created a new one - issue_297

To test the logger, I guess, you can do this, from within the `client`:

```
    log::info!("testing it....");
```

The log files will be created in the subdirectory `logs/`

**Comment by MathJud** [comment]
fixed by #346
